libpdbg/device.c

```diff
@@ -48,11 +52,78 @@ static const char *take_name(const char *name)
 	return name;
 }
 
+/*
+ * If a string contains alphanumeric characters then it extracts
+ * the number from that string (from the end only) and
+ * returns the extracted number and the chopped off string
+ */ 
+static unsigned int strip_num_at_the_end(char *str) 
+{
+	unsigned int only_number;
+	int len = strlen(str);
+	char new_num_str[len];
+
+	int i;
+	/*Traverse from the end of the string looking for alphabet after the digit */
+	for(i = (len - 1); i >= 0; i--) {
+		if(!isdigit(str[i])) {
+		/* This is the first character from the end. So 
+		 * copy the number starting from the next index 
+		 */
+		strcpy(new_num_str, &str[i+1]);
+		break;
+		}
+	}
+	str[i+1] = '\0';
+	only_number = atoi(new_num_str);
+	return only_number;
+}
+
+/*
+ * Find the class the target belongs to, and add the target to
+ * the list at a specific position so that it remains sorted
+ */ 
+void sort_and_add_class_link(struct pdbg_target *target)
+{
+	struct pdbg_target_class *target_class;
+	struct pdbg_target *node = NULL;
+
+	target_class = get_target_class(target);
+
+	if (list_empty(&target_class->targets)) {
+		list_add(&target_class->targets, &target->class_link);
+		return;
+	}
+
+	bool add_before = false;
+	pdbg_for_each_target(target_class->name, NULL, node) {
+		int cmp = dt_cmp_subnodes(node, target);
+		if (cmp == 0) {
+			/* Duplicate entry, don't add */
+			return;
+		}
+		
+		/* insert before the first node that's larger
+		 * the the node we're inserting 
+		 */
+		if (cmp > 0){
+			add_before = true;
+			break;
+		}
+	}
+
+	if(add_before) {
```

Comment:
Combine these two condition in `if-else` statement and remove `return` from the if. More code clarity